llvm/test/CodeGen/X86/tls-desc.ll

```diff
@@ -0,0 +1,165 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -mtriple=i686-unknown-unknown --relocation-model=pic -enable-tlsdesc | FileCheck %s --check-prefix=X86
+; RUN: llc < %s -mtriple=x86_64-pc-linux-gnux32 --relocation-model=pic -enable-tlsdesc | FileCheck %s --check-prefix=X32
+; RUN: llc < %s -mtriple=x86_64-unknown-unknown --relocation-model=pic -enable-tlsdesc | FileCheck %s --check-prefix=X64
+
+@x = thread_local global i32 0, align 4
+@y = internal thread_local global i32 0, align 4
```

Comment:
GCC's x86-64 port assumes that FLAGS_REG and RAX are changed while all other registers are preserved.

Ideally, show that some registers are not clobbered.

Reply:
Thanks for the info! In that way, we need to define a new pseudo instruction that only clobbers the two register.
I'll try to do it in the next reversion.

Reply:
Done.

Reply:
Thanks for the update. We need at least 2 `internal thread_local` to trigger general dynamic => local dynamic optimization, and demonstrate that TLSDESC optimizes

Reply:
Done.

I don't see GCC generate `leaq    _TLS_MODULE_BASE_@TLSDESC(%rip), %rax` but optimized them away https://godbolt.org/z/Ksv43bsq6

Reply:
That example allows the variables to be optimized out entirely because there's no way they can ever be modified. https://godbolt.org/z/63oz4xrj1 shows an example where it's generating LD model accesses.

Note that in some cases I've seen GCC generate a local anchor symbol instead of using `_TLS_MODULE_BASE_`. That is less optimal than using `_TLS_MODULE_BASE_`, because multiple uses of `_TLS_MODULE_BASE_` across multiple TUs can be resolved by the linker to a single GOT slot (pair), whereas a local anchor symbol will always have its own GOT slot (pair) that won't be shared with any other TU.

Reply:
Thanks @frobtech for the explanation. IIUC, GCC will use `_TLS_MODULE_BASE_` if there are more than 1 local TLS variables, right?

I have no idea how to do it smartly in codegen in LLVM since we lower it statelessly. Maybe let's land this first? @MaskRay @frobtech

Reply:
Done.

Reply:
I think it's worthwhile to have test cases that are `dso_local` but not internal linkage as well, and mixing the two.  These can and should also use `_TLS_MODULE_BASE_` with the secondary `dtpoff` offset after the TLSDESC call returns. e.g.
```
extern __thread int x __attribute__((visibility("hidden")));
static __thread int y;
int foo() { return ++x + ++y; }
```

The ideal thing is probably *not* to use `_TLS_MODULE_BASE_` when its result is used only once in the function, regardless of the linkage details of the referenced TLS symbol.  That's because a second addition with the `dtpoff` value will always be needed for that. If it's a case where that addition is "free", e.g. because it can be rolled into the displacement of a load you're doing anyway, then using `_TLS_MODULE_BASE_` is ideal even if it's the only use in the function (because the GOT slot & relocs will be shared with any other such use in the whole link).  However, if it requires an extra instruction to do the add after the TLSDESC call returns, then it's probably better overall to have the separate reloc for just that access.

Reply:
Added in `f5`.

> The ideal thing is probably not to use `_TLS_MODULE_BASE_` when its result is used only once in the function

I think it is demonstrated by `f3`?